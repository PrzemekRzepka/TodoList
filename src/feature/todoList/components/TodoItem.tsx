import { View, Button, Text, StyleSheet } from "react-native"

interface TodoItemProps {
    id: string,
    text: string,
    onRemove: (id: string) => void
}


export default function TodoItem({ id, text, onRemove }: TodoItemProps) {
    return (
        <View style={style.constainer}>
            <View style={{ flex: 1, padding: 20, alignItems: 'flex-start' }}>
                <Text>{new Date(Date.now()).toString()} :</Text>
                <Text>{text}</Text>
            </View>
            <Button title="Remove item" onPress={() => onRemove(id)} />
        </View>
    )

}

const style = StyleSheet.create({
    constainer: {
        borderRadius: 0,
        borderWidth: 1,
        marginVertical: 10,
        padding: 6,
    },
    button: {

    }
})