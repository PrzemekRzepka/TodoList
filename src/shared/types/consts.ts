export const REMOVE_ITEM_TEXT = "X";
export const ADD_TEXT = "ADD";
export const TEXT_INPUT_PLACEHOLDER_TEXT = "To do task text";