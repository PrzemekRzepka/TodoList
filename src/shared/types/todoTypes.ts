export type TodoItem = {
    id: string,
    text: string,
    isDone: boolean,
    createdAt: string
}

export type TodoList = TodoItem[];