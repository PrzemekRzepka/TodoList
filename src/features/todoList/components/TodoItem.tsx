import { View, Text, StyleSheet } from "react-native"
import { AdvancedCheckbox } from 'react-native-advanced-checkbox';

import { REMOVE_ITEM_TEXT } from "../../../shared/types/consts"

import Button from "./Button"

interface TodoItemProps {
    id: string,
    text: string,
    isDone: boolean,
    createdAt: string,
    onRemove: (id: string) => void,
    onDone: (id: string) => void
}


export default function TodoItem({ id, text, isDone, createdAt, onRemove, onDone }: TodoItemProps) {
    const styles = getStyles(isDone)

    return (
        <View style={styles.constainer}>
            <AdvancedCheckbox
                value={isDone}
                onValueChange={() => onDone(id)}
            />
            <View style={styles.textContainer}>
                <Text style={styles.text}>{createdAt}</Text>
                <Text style={[styles.text, { fontSize: 20 }]}>{text}</Text>
            </View>
            <Button style={styles.button} title={REMOVE_ITEM_TEXT} onPress={() => onRemove(id)} />
        </View>
    )

}

const getStyles = (isDone: boolean) => StyleSheet.create({
    constainer: {
        flexDirection: 'row',
        borderRadius: 10,
        borderWidth: 1,
        marginVertical: 4,
        paddingHorizontal: 16,
        backgroundColor: isDone ? '#5a6069' : '#3b598a',
        justifyContent: 'center',
        alignItems: 'center'
    },
    textContainer: {
        flex: 1,
        padding: 20,
        alignItems: 'flex-start'

    },
    text: {
        color: isDone ? 'gray' : 'white'
    },
    button: {
        width: 40,
        height: 40,

    }
})