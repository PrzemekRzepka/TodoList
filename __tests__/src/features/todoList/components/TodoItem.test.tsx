import { render, waitFor, fireEvent } from "@testing-library/react-native"
import TodoItem from "../../../../../src/features/todoList/components/TodoItem"
import { REMOVE_ITEM_TEXT } from "../../../../../src/shared/types/consts";

const TEST_PROPS = {
    id: 'testId',
    text: 'testText',
    createdAt: '20.10.1999',
    isDone: false,
    onDone: jest.fn(),
    onRemove: jest.fn()
}

describe('TodoItem component', () => {
    describe('snapshots', () => {
        it('snapshot test', () => {
            const snapshot = render(<TodoItem {...TEST_PROPS} />)
            expect(snapshot).toMatchSnapshot()
        })
    })

    describe('properties', () => {
        it('given props and mocked date should render mocked date', async () => {

            const { getByText } = render(<TodoItem {...TEST_PROPS} />)

            await waitFor(() => {
                expect(getByText(TEST_PROPS.createdAt)).toBeTruthy();
            })
        })

        it('given props should render text', async () => {
            const { getByText } = render(<TodoItem {...TEST_PROPS} />)

            await waitFor(() => {
                expect(getByText(TEST_PROPS.text)).toBeTruthy();
            })
        })

        it('given props should call onRemove after click', async () => {
            const { getByText } = render(<TodoItem {...TEST_PROPS} />)

            await waitFor(() => {
                const button = getByText(REMOVE_ITEM_TEXT);
                fireEvent.press(button);
            })

            expect(TEST_PROPS.onRemove).toHaveBeenCalled()
        })


    })
})