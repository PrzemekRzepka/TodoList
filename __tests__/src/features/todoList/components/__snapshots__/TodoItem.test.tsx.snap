// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem component snapshots snapshot test 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#3b598a",
      "borderRadius": 10,
      "borderWidth": 1,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginVertical": 4,
      "paddingHorizontal": 16,
    }
  }
>
  <View
    accessibilityHint="Toggles checkbox to on"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginVertical": 8,
        },
        undefined,
        {
          "opacity": 1,
        },
      ]
    }
  >
    <View
      collapsable={false}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "#ccc",
          "borderRadius": 4,
          "borderWidth": 1.5,
          "height": 24,
          "justifyContent": "center",
          "opacity": 1,
          "transform": [
            {
              "scale": 1,
            },
            {
              "rotate": "0deg",
            },
          ],
          "width": 24,
        }
      }
    />
  </View>
  <View
    style={
      {
        "alignItems": "flex-start",
        "flex": 1,
        "padding": 20,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
        }
      }
    >
      20.10.1999
    </Text>
    <Text
      style={
        [
          {
            "color": "white",
          },
          {
            "fontSize": 20,
          },
        ]
      }
    >
      testText
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#367ff5",
        "borderRadius": 10,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
        }
      }
    >
      X
    </Text>
  </View>
</View>
`;
